memcached.c

```diff
@@ -8405,11 +8764,25 @@ int main (int argc, char **argv) {
     uriencode_init();
 
     /* enter the event loop */
-    if (event_base_loop(main_base, 0) != 0) {
-        retval = EXIT_FAILURE;
+    while (!stop_main_loop) {
+        if (event_base_loop(main_base, EVLOOP_ONCE) != 0) {
+            retval = EXIT_FAILURE;
+            break;
+        }
     }
 
-    stop_assoc_maintenance_thread();
+    fprintf(stderr, "Gracefully stopping\n");
+    stop_threads();
+    int i;
+    // FIXME: make a function callable from threads.c
+    for (i = 0; i < max_fds; i++) {
+        if (conns[i] && conns[i]->state != conn_closed) {
+            conn_close(conns[i]);
```

Comment:
nit: `conn_free()` as well?

Reply:
That's funny... conn_free() is only ever called if a malloc fails during conn_new... so it's probably never actually been executed before. We should do this but with more deliberate testing to avoid accidentally adding a crash bug to the stop routine. I'm also changing the way connection memory works anyway.